EIPS/eip-7002.md

```diff
@@ -0,0 +1,372 @@
+    return trie.root
+
+assert block.header.exits_root == compute_trie_root_from_indexed_data(block.body.exits)
+```
+
+2. The list of exit operations contained in the block body **MUST** be equivalent to list of exits at the head of the exit precompile's exit message queue up to the maximum of `MAX_EXITS_PER_BLOCK`, respecting the order in the queue. To illustrate:
+
+```python
+class Exit(object):
+    source_address: Bytes20
+    validator_index: uint64
+
+queue_head_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_HEAD_STORAGE_SLOT)
+queue_tail_index = sload(EXIT_PRECOMPILE_ADDRESS, EXIT_MESSAGE_QUEUE_TAIL_STORAGE_SLOT)
+num_exits_in_queue = queue_tail_index - queue_head_index
+num_exits_to_dequeue = min(num_exits_in_queue, MAX_EXITS_PER_BLOCK)
+
+# Retrieve exits from the queue
+expected_exits = []
+for i in range(num_exits_to_dequeue):
+    queue_storage_slot = EXIT_MESSAGE_QUEUE_STORAGE_OFFSET + (queue_head_index + i) * 2
+    source_address = SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot)
+    validator_index = SLOAD(EXIT_PRECOMPILE_ADDRESS, queue_storage_slot + 1)
+    exit = Exit(
+        source_address=Bytes20(source_address),
+        validator_index=uint64(validator_index),
+    )
+    expected_exits.append(exit)
+
+# Compare retrieved exits to the list in the block body
+assert block.body.exits == expected_exits
+```
+
+A block that does not satisfy the above conditions **MUST** be deemed invalid.
+
+### Consensus layer
+
+In progress.
+
+Sketch of spec:
+
+* New operation `ExecutionLayerExit`
+* Will show up in `ExecutionPayload` as a vector bound by length `MAX_EXITS_PER_BLOCK`
+* New function in `process_execution_layer_exit` that has similar functionality to `process_voluntary_exit` but that can fail validations (e.g. validator is already exited) without the block failing (similar to deposit coming from EL)
+* `process_execution_layer_exit` called in `process_operations` for each `ExecutionLayerExit` found in the `ExecutionPayload`
+
+## Rationale
+
+### Stateful precompile
+
+This specification utilizes a *stateful* precompile for simplicity and future-proofness. While precompiles are a well-known quantity, none to date have associated EVM state at the address.
+
+The alternative designs are (1) to utilize a precompile or opcode for the functionality and write a separate specified space in the EVM -- e.g. `0xFF..FF` -- or (2) to place the required state into the block and require the previous block header as an input into the state transition function (e.g. like 1559 `base_fee`).
+
+Alternative design (1) is essentially using a stateful precompile but dissociating the state into a separate address. At first glance, this split appears unnecessarily convoluted when we could store the location of the `CALL` and the associated state in the same address. That said, there might be unexpected engineering constraints around precompiles in existing clients that make this a preferable path.
```

Comment:
One more consideration here is that it'd be nice to be able to know the current fee in a smart contract, so it'd be nice to have a stateful precompile with view functions for state. That would make it pretty complex for a precompile.

I think a prerompile + separate address for storage and view functions is not a bad option.

Reply:
We were discussing having two precompiles:

1. A precompile where to perform the exit (reads and writes to (2))
2. A precompile with a fee-read that has all of the state utilized by (1)

I think this reasonable. CC: @adietrichs for input

Reply:
Might have been answered before...when precompile vs when opcode?